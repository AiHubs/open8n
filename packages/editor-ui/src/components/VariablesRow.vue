<script lang="ts" setup>
import { computed, onMounted, PropType, ref } from 'vue';
import { EnvironmentVariable } from '@/Interface';
import { useI18n } from '@/composables';

const i18n = useI18n();

const emit = defineEmits(['save', 'cancel', 'edit', 'delete']);

const props = defineProps({
	data: {
		type: Object as PropType<EnvironmentVariable>,
		default: () => ({}),
	},
	editing: {
		type: Boolean,
		default: false,
	},
});

const modelValue = ref<EnvironmentVariable>(props.data);

const formValid = computed(() => {
	return modelValue.value.key !== '';
});

const keyInputRef = ref<HTMLElement>();
const valueInputRef = ref<HTMLElement>();

onMounted(() => {
	keyInputRef.value?.focus();
});

async function onCancel() {
	emit('cancel', modelValue.value);
}

async function onSave() {
	emit('save', modelValue.value);
}

async function onEdit() {
	emit('edit', modelValue.value);
}

async function onDelete() {
	emit('delete', modelValue.value);
}
</script>

<template>
	<tr :class="$style.variablesRow">
		<td>
			<span v-if="!editing">{{ data.key }}</span>
			<n8n-input
				v-else
				data-testid="variable-row-key-input"
				:required="true"
				:placeholder="i18n.baseText('variables.editing.key.placeholder')"
				v-model="modelValue.key"
				ref="keyInputRef"
			/>
		</td>
		<td>
			<span v-if="!editing">{{ data.value }}</span>
			<n8n-input
				v-else
				data-testid="variable-row-value-input"
				:placeholder="i18n.baseText('variables.editing.value.placeholder')"
				v-model="modelValue.value"
				ref="valueInputRef"
			/>
		</td>
		<td class="text-right">
			<div v-if="editing">
				<n8n-button
					data-testid="variable-row-cancel-button"
					type="tertiary"
					class="mr-xs"
					@click="onCancel"
				>
					{{ i18n.baseText('variables.row.button.cancel') }}
				</n8n-button>
				<n8n-button
					data-testid="variable-row-save-button"
					:disabled="!formValid"
					type="primary"
					@click="onSave"
				>
					{{ i18n.baseText('variables.row.button.save') }}
				</n8n-button>
			</div>
			<div :class="$style.hoverButtons" v-else>
				<n8n-button
					data-testid="variable-row-edit-button"
					type="tertiary"
					class="mr-xs"
					@click="onEdit"
				>
					{{ i18n.baseText('variables.row.button.edit') }}
				</n8n-button>
				<n8n-button
					data-testid="variable-row-delete-button"
					type="tertiary"
					class="mr-xs"
					@click="onDelete"
				>
					{{ i18n.baseText('variables.row.button.delete') }}
				</n8n-button>
			</div>
		</td>
	</tr>
</template>

<style lang="scss" module>
.variablesRow {
	&:hover {
		.hoverButtons {
			opacity: 1;
		}
	}
}

.hoverButtons {
	opacity: 0;
	transition: opacity 0.2s ease;
}
</style>
