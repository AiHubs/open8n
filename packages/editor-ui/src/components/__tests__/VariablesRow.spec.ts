import VariablesRow from '../VariablesRow.vue';
import { EnvironmentVariable } from '@/Interface';
import { fireEvent, getByTestId, render } from '@testing-library/vue';

describe('VariablesRow', () => {
	const environmentVariable: EnvironmentVariable = {
		id: 1,
		key: 'key',
		value: 'value',
	};

	it('should render correctly', () => {
		const wrapper = render(VariablesRow, {
			props: {
				data: environmentVariable,
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
		expect(wrapper.container.querySelectorAll('td')).toHaveLength(3);
	});

	it('should show edit and delete buttons on hover', async () => {
		const wrapper = render(VariablesRow, {
			props: {
				data: environmentVariable,
			},
		});

		await fireEvent.mouseEnter(wrapper.container);

		expect(wrapper.getByTestId('variable-row-edit-button')).toBeVisible();
		expect(wrapper.getByTestId('variable-row-delete-button')).toBeVisible();
	});

	it('should show key and value inputs in edit mode', async () => {
		const wrapper = render(VariablesRow, {
			props: {
				data: environmentVariable,
				editing: true,
			},
		});

		await fireEvent.mouseEnter(wrapper.container);

		expect(wrapper.getByTestId('variable-row-key-input')).toBeVisible();
		expect(wrapper.getByTestId('variable-row-key-input').querySelector('input')).toHaveValue(
			environmentVariable.key,
		);
		expect(wrapper.getByTestId('variable-row-value-input')).toBeVisible();
		expect(wrapper.getByTestId('variable-row-value-input').querySelector('input')).toHaveValue(
			environmentVariable.value,
		);
	});

	it('should show cancel and save buttons in edit mode', async () => {
		const wrapper = render(VariablesRow, {
			props: {
				data: environmentVariable,
				editing: true,
			},
		});

		await fireEvent.mouseEnter(wrapper.container);

		expect(wrapper.getByTestId('variable-row-cancel-button')).toBeVisible();
		expect(wrapper.getByTestId('variable-row-save-button')).toBeVisible();
	});
});
