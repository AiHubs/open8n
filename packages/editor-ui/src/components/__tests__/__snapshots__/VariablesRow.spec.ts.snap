// Vitest Snapshot v1

exports[`VariablesRow > should render correctly 1`] = `
"<tr class=\\"variablesRow\\">
  <td><span>key</span></td>
  <td><span>value</span></td>
  <td class=\\"variables-usage-column\\">
    <n8n-tooltip-stub justifybuttons=\\"flex-end\\" buttons=\\"\\" placement=\\"top\\"><span class=\\"usageSyntax\\">$vars.key</span></n8n-tooltip-stub>
  </td>
  <td>
    <div class=\\"buttons hoverButtons\\"><button aria-disabled=\\"false\\" aria-busy=\\"false\\" aria-live=\\"polite\\" class=\\"mr-xs button button tertiary medium\\" data-test-id=\\"variable-row-edit-button\\">
        <!----><span> Edit </span></button><button aria-disabled=\\"false\\" aria-busy=\\"false\\" aria-live=\\"polite\\" class=\\"mr-xs button button tertiary medium\\" data-test-id=\\"variable-row-delete-button\\">
        <!----><span> Delete </span></button></div>
  </td>
</tr>"
`;

exports[`VariablesRow > should show key and value inputs in edit mode 1`] = `
"<tr class=\\"variablesRow\\">
  <td>
    <div class=\\"el-input el-input--large n8n-input\\" data-test-id=\\"variable-row-key-input\\" required=\\"required\\">
      <!----><input type=\\"text\\" autocomplete=\\"off\\" placeholder=\\"Enter a name\\" class=\\"el-input__inner\\">
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </td>
  <td>
    <div class=\\"el-input el-input--large n8n-input\\" data-test-id=\\"variable-row-value-input\\">
      <!----><input type=\\"text\\" autocomplete=\\"off\\" placeholder=\\"Enter a value\\" class=\\"el-input__inner\\">
      <!---->
      <!---->
      <!---->
      <!---->
    </div>
  </td>
  <td class=\\"variables-usage-column\\">
    <n8n-tooltip-stub justifybuttons=\\"flex-end\\" buttons=\\"\\" placement=\\"top\\"><span class=\\"usageSyntax\\">$vars.key</span></n8n-tooltip-stub>
  </td>
  <td>
    <div class=\\"buttons\\"><button aria-disabled=\\"false\\" aria-busy=\\"false\\" aria-live=\\"polite\\" class=\\"mr-xs button button tertiary medium\\" data-test-id=\\"variable-row-cancel-button\\">
        <!----><span> Cancel </span></button><button aria-disabled=\\"false\\" aria-busy=\\"false\\" aria-live=\\"polite\\" class=\\"button button primary medium\\" data-test-id=\\"variable-row-save-button\\">
        <!----><span> Save </span></button></div>
  </td>
</tr>"
`;
